### DISCUSSION TITLE
Setting up tabbed diagrams

### DISCUSSION THREAD

**Discussion by X39** [Blazor-Diagrams/Blazor.Diagrams#190] [Q&A] [answered]
Hello there,

i am currently struggling getting the Diagrams component working in a multi-tabbed system but struggling to get it working.

My main assumption here was that `Diagram` was actually responsible for handling the nodes and links but it seems as if this is not true according to my results (which shows the same diagram with each tab) but it seems like that is not the case in my testing.

Full sources are available here: https://github.com/X39/X39.Systems.ServiceOrchestrator

thanks for any help,
X39

**Comment by zHaytam** [comment]
Hello, I'll try your project right now and see.

**Comment by zHaytam** [comment]
![image](https://user-images.githubusercontent.com/34218324/166100754-0ef4e99d-9d9e-4515-827a-7ac9cbfe5787.png)

Not sure where can I see the tabs. Can you give me instructions in order to run the project please?

**Reply by X39** [reply]
You have to double click a service string in the list😅😂

It is kinda confusing still due to wip

**Comment by zHaytam** [comment]
Hello, I see that you're creating a single `Diagram` instance and using it for all tabs. That's the problem. You need one diagram per tab, otherwise, they will share the links/nodes and you'll get a lot of cross-interactions.

```
public Diagram Diagram { get; } = new(new DiagramOptions
    {
        DeleteKey = "Delete",
        AllowMultiSelection = true,
        AllowPanning = true,
        GridSize = GridSize,
        Links = new DiagramLinkOptions
        {
            DefaultPathGenerator = PathGenerators.Straight,
            DefaultRouter = Routers.Orthogonal,
        },
        Zoom = new DiagramZoomOptions
        {
            Minimum = 0.125,
            Maximum = 1,
            Inverse = true,
        },
    });
```

**Reply by X39** [reply]
Hey, many thanks for your time first of all.

Just to clear up a possible missunderstanding on my side here some insight on how the tabs are created:

- in `Designer/PageParts/ToolBoxes/ServiceToolBox.razor#L53-57` a new `ServiceDataContext` is created and added to the tabs list with `Designer/PageParts/DesignView/DesignView.razor` being the display component
- The `Diagram` is created in that `ServiceDataContext`
- Each tab is backed by its very own `ServiceDataContext`
- Tabs are displayed in `Designer/Components/TabControl/TabControl.razor` using `DynamicComponent`
- `DynamicComponent` sets the `ServiceDataContext` of the `DesignView`



Is this actually reusing the same instance? because the other properties are actually properly altered.

It may be a lot to ask but i would like to invite you to show me where i am wrong here 😅 because i am quite lost with what is wrong.

**Comment by zHaytam** [comment, marked as answer]
![](https://i.imgur.com/vFnHNCY.gif)

The design view was not getting re-rendered because it can't figure out that it changed. Consider adding `@key`:
```
<SplitContainer SplitDirection="SplitContainer.ESplitDirection.Horizontal" InitialSplit="0.75 0.25">
    <ChildA>
        <DesignView @key="DataContext" DataContext="DataContext"/>
    </ChildA>
    <ChildB>
        <PropertyView/>
    </ChildB>
</SplitContainer>
```

**Reply by X39** [reply]
Many Thanks!
Especially for the effort you went through to help me out :3
